packages/context/src/inject.ts

```diff
@@ -5,14 +5,26 @@
 
 import * as assert from 'assert';
 import { Reflector } from './reflect';
-import {BoundValue} from './binding';
+import { BoundValue, ValueOrPromise } from './binding';
+import { Context } from './context';
 
 const REFLECTION_CDI_KEY = 'loopback:inject:constructor';
 const REFLECTION_PDI_KEY = 'loopback:inject:properties';
 
+/**
+ * A function to provide resolution of injected values
+ */
+export interface ResolverFunction {
+  (ctx: Context, injection: Injection): ValueOrPromise<BoundValue>;
+}
+
+/**
+ * Descriptor for an injection point
+ */
 export interface Injection {
-  bindingKey: string;
-  metadata?: { [attribute: string]: BoundValue; };
+  bindingKey: string; // Binding key
+  metadata?: { [attribute: string]: BoundValue; }; // Related metadata
+  resolve?: ResolverFunction; // A custom resolve function
```

Comment:
nit: comments unnecessary (at least bindingKey // Binding Key), they will be described in the places where the actual injection is used